components/script/dom/webglbuffer.rs

```diff
@@ -23,30 +24,111 @@ fn target_is_copy_buffer(target: u32) -> bool {
         target == WebGL2RenderingContextConstants::COPY_WRITE_BUFFER
 }
 
+#[derive(JSTraceable, MallocSizeOf)]
+struct DroppableWebGLBuffer {
+    #[no_trace]
+    id: WebGLBufferId,
+    marked_for_deletion: Cell<bool>,
+    attached_counter: Cell<u32>,
+    context: WeakRef<WebGLRenderingContext>,
+}
+
+impl DroppableWebGLBuffer {
+    pub(crate) fn new(
+        id: WebGLBufferId,
+        marked_for_deletion: Cell<bool>,
+        attached_counter: Cell<u32>,
+        context: WeakRef<WebGLRenderingContext>,
+    ) -> Self {
+        Self {
+            id,
+            marked_for_deletion,
+            attached_counter,
+            context,
+        }
+    }
+}
+
+impl DroppableWebGLBuffer {
+    pub(crate) fn is_marked_for_deletion(&self) -> bool {
+        self.marked_for_deletion.get()
+    }
+
+    pub(crate) fn set_marked_for_deletion(&self, marked_for_deletion: bool) {
+        self.marked_for_deletion.set(marked_for_deletion);
+    }
+
+    pub(crate) fn get_attached_counter(&self) -> u32 {
+        self.attached_counter.get()
+    }
+
+    pub(crate) fn set_attached_counter(&self, attached_counter: u32) {
+        self.attached_counter.set(attached_counter);
+    }
+
+    pub(crate) fn id(&self) -> WebGLBufferId {
+        self.id
+    }
+
+    pub(crate) fn is_attached(&self) -> bool {
+        self.get_attached_counter() != 0
+    }
+
+    pub(crate) fn is_deleted(&self) -> bool {
+        self.is_marked_for_deletion() && !self.is_attached()
+    }
+
+    pub(crate) fn delete(&self, operation_fallibility: Operation) {
+        assert!(self.is_deleted());
+        let context = self.context.root().unwrap();
```

Comment:
Instead of panicking, we should warn here that we're leaking a webgl buffer resource and then return.

Reply:
I was actually referring to the unwrap call on this line.

Reply:
what do you mean?

Reply:
I think I found a solution!